### PR TITLE
Add arguments_size and visit_arguments for std::tuple and pybind11::tuple

### PR DESCRIPTION
**Pull request by andrei-sa**
The functions have to be defined for Arguments rather than Sample because Model is supposed to take Arguments as an argument (when predicting).
The unit tests demonstrate use cases necessary in Model when adapting Samples to singly typed Algorithms.

### CODE REVIEW

File: Python/Include/Sample.hpp [hunk 1]
```diff
@@ -10,27 +10,41 @@ namespace Rover {
...
   };
 
+  //! Applies a function to every element of a Python tuple
+  /*!
+    \param func Function taking a tuple element and its index as arguments.
+    \param tuple Python tuple of the arguments.
+  */
+  template<typename Func>
+  void visit_arguments(Func&& func, pybind11::tuple& tuple) {
+    for(std::size_t i = 0LU; i < tuple.size(); ++i) {
```
Comment (kranar) [inline Python/Include/Sample.hpp]: ```auto i = std::size_t(0)```

File: Include/Rover/TupleArguments.hpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  template<typename Func, typename... A>
+  void visit_arguments(Func&& func, std::tuple<A...>& tuple) {
+    std::apply([&](auto&&... args) {
+      auto i = 0LU;
+      (func(args, i++), ...);
+    }, tuple);
+  }
+
+  //! Returns the number of elements in a tuple
+  template<typename... A>
```
Comment (kranar) [inline Include/Rover/TupleArguments.hpp]: Place these definitions in ```Sample.hpp``` instead of needing to include another file to get this to work.

**Review by kranar — CHANGES_REQUESTED** [review]
Looks amazing.

**Comment by andrei-sa** [comment]
This is actually the part I didn't like about the "always auto" convention. size_t aliases a built-in type, so size_t(0) is a mere C cast. That violates the "always use the least powerful cast" convention and, in order to respect it, I need to code it as auto i = static_cast<size_t>(0), which is too verbose. size_t i = 0 is in turn an implicit cast, which is just as powerful as static_cast. One just have to decide which convention is more important for them, or use the wordy static_cast<size_t>(0) syntax (or create a macro shortcut for it).

**Review by kranar — APPROVED** [review]

**Comment by kranar** [comment]
This is great.

**Pull request merged**